quark/foreach.py

```diff
@@ -0,0 +1,66 @@
+# -*- coding: utf-8 *--
+
+import os
+import subprocess
+from argparse import ArgumentParser
+
+from quark.utils import DirectoryContext
+
+from .subproject import Subproject
+
+
+def run():
+    parser = ArgumentParser(
+        description="""
+        Evaluates an arbitrary shell command in each submodule.
+        The command has access to the variables $name, $sm_path, $displaypath,
+        $sha1, $toplevel, $rev, $version_control:
+        $name is the name of the submodule;
+        $sm_path is the path of the submodule relative to the superproject;
+        $displaypath is the path of the submodule relative to the root
+        directory;
+        $version_control is the version control used by the subproject
+        (git/svn);
+        $sha1 is the commit of the subproject ( empty string if it is a
+        svn repository );
+        $rev is the revision of the subproject ( empty string if it is a
+        git repository );
+        $toplevel is the absolute path to the top-level of the
+        immediate superproject.
+        """
+    )
+    parser.add_argument(
+        "-q", "--quiet",
+        action="store_true",
+        help="Only print error messages"
+    )
+    parser.add_argument(
+        "command",
+        action="store",
+        nargs="+",
+        help="The command that will be run for every dependency"
+    )
+
+    optlist = parser.parse_args()
+
+    root, modules = Subproject.create_dependency_tree(
+        os.getcwd(), update=False
+    )
+
+    for path, module in modules.items():
+        cmd_env = dict(os.environ)
+        cmd_env.update(module.get_env_variables(toplevel=os.getcwd()))
+
+        with DirectoryContext(module.directory):
+            if not optlist.quiet:
+                print("Entering {}".format(cmd_env["name"]))
+
+            subprocess.check_call(
+                optlist.command[0],
```

Comment:
Any reason why this should be `optlist.command[0]` rather than `optlist.command`?

Reply:
I checked, unfortunately with `shell = True` it's not so simple, as I remembered this morning.

`shell = True` with a list does the equivalent of `check_call(['/bin/sh', '-c'] + commands, ...)`, which is _not_ telling the shell to execute `commands[0]` using `commands[1:]` as arguments (as one would expect). Instead, the first argument to `-c` is the whole command line that the shell has to execute (i.e. "the body of the script"). The arguments that follow are interpreted just as the `argv` of this script, so:

```
In [11]: subprocess.check_call(["echo", "hello", "world"], shell=True)

Out[11]: 0
```

prints nothing, as the command line being executed is just `echo`; the other arguments are the arguments _of the "temporary" script_, as can be seen:

```
In [12]: subprocess.check_call(["echo $0 - $1", "hello", "world"], shell=True)
hello - world
Out[12]: 0

```

The only way I see to forward the already tokenized arguments (barring getting into the tunnel of escaping, which I would avoid completely) is to do something as hideous as

```
subprocess.check_call(["$@", "placeholder"] + optlist.command, shell = True)
```

However this breaks the case of the whole script to be `eval`-ed in a single string. `git submodules foreach` (which is actually a huge `/bin/sh` script itself) actually "solves" this by doing

        if test $# -eq 1
        then
          eval "$1"
        else
          "$@"
        fi

so, if there's a single argument do an `eval`, otherwise consider all the arguments as a pre-tokenized command line and execute it accordingly. The equivalent in our case would be something like:

```
if len(optlist.command) == 1:
    cmd = optlist.command
else:
    cmd = ['$@', "_dummy_"] + optlist.command
subprocess.check_call(cmd, shell = True)
```

I'm not sure if I like this behavior, but if we are trying to match `git submodule foreach` that's what should be implemented.

---

So, to recap, we have essentially three alternatives:

- do as we are doing now; only the first argument is considered, and it's the whole string to be `eval`-ed by `/bin/sh`; I'd only add a check to quit immediately with an error if more than one argument is passed, to avoid mistakes on user's part, expecting that multiple arguments are actually employed;
- always do the `"$@"` dance, so stuff is always executed as tokenized by the shell that invoke `quark foreach`; this gives me some chills about whether this trick always works (how many arguments are actually supported? does stuff break if it's not a single command but a more complex script, ...) and requires extra, separate support for Windows (where this whole thing should be bypassed? who knows?), and breaks the "I want to pass whole script in a single argument" case
- do the `"$@"` dance _but_ special-cased for a single argument, as `git subproject foreach` does; this is ugly, has all the drawbacks of the previous bullet but is coherent with `git submodule foreach`.

Personally, I'd go with the first option. There's nothing special to do with Windows, it's already implemented and the only extra thing to do is to add a check to crash with a helpful error message in the multiple arguments case. The other options look to me brittle kludges, for no big gain; if you struggle too much with escaping you are probably doing something too complex anyway for a one-liner, just rewrite it as a proper script.

Reply:
Ok, fixed in 31f587d